File: frontend/src/types/screening.ts
// Mirrors backend/app/schemas/screening.py and the preprocess/health endpoints.
// Keep these in sync manually — there's no shared codegen step yet.

export interface OCRField {
  value: string;
  confidence: number; // 0.0–1.0
}

export interface OCRResult {
  fields: Record<string, OCRField>;
  mean_confidence: number;
}

export interface ELARegion {
  x: number;
  y: number;
  width: number;
  height: number;
  mean_error: number;
}

export interface ELAResult {
  tamper_score: number; // 0.0–1.0, higher = more evidence of splicing
  heatmap_path: string;
  compression_warning: boolean;
  forensic_notes: string;
  suspicious_regions: ELARegion[];
}

export interface FaceMatchResult {
  similarity_score: number; // 0.0–1.0
  face_detected_on_id: boolean;
  face_detected_on_selfie: boolean;
}

export type Verdict = "Genuine" | "Suspicious" | "Fake";

export interface ScreeningVerdict {
  trust_score: number; // 0–100
  verdict: Verdict;
  explanation: string[];
}

export interface ScreeningResponse {
  document_id: string;
  ocr: OCRResult;
  ela: ELAResult;
  face_match: FaceMatchResult;
  result: ScreeningVerdict;
}

export interface PreprocessResponse {
  document_id: string;
  processed_image_path: string;
  crop_method: "perspective_warp" | "bounding_box" | "none";
  corners_detected: number[][] | null;
}

export interface HealthResponse {
  status: string;
}

// --- Local UI state (not backend-mirrored) --------------------------------

export interface AuditEntry {
  id: string;
  timestamp: string; // ISO
  fileName: string;
  trustScore: number;
  verdict: Verdict;
  documentId: string;
}

File: backend/app/schemas/screening.py
"""
Response/request contracts for the document screening pipeline.
Kept separate from ORM models so the API contract is stable even if
persistence changes.
"""
from typing import Literal
from pydantic import BaseModel, Field


class OCRField(BaseModel):
    value: str
    confidence: float = Field(ge=0.0, le=1.0)


class OCRResult(BaseModel):
    fields: dict[str, OCRField]
    mean_confidence: float = Field(ge=0.0, le=1.0)
    # Sensitive numeric IDs are never populated with realistic digits in this
    # codebase's sample/test payloads — see OCR module docstring.


class ELAResult(BaseModel):
    tamper_score: float = Field(ge=0.0, le=1.0, description="0 = no evidence of splicing, 1 = strong evidence")
    heatmap_path: str
    compression_warning: bool = Field(
        default=False,
        description="True if background noise floor suggests heavy/repeated JPEG re-compression "
                    "(e.g. messaging-app forwards), which can inflate tamper_score independent of forgery.",
    )
    forensic_notes: str = Field(
        default="",
        description="Human-readable caveat for the officer UI, populated when compression_warning is True.",
    )
    suspicious_regions: list[dict] = Field(default_factory=list)


class FaceMatchResult(BaseModel):
    similarity_score: float = Field(ge=0.0, le=1.0)
    face_detected_on_id: bool
    face_detected_on_selfie: bool


class ScreeningVerdict(BaseModel):
    trust_score: int = Field(ge=0, le=100)
    verdict: Literal["Genuine", "Suspicious", "Fake"]
    explanation: list[str] = Field(default_factory=list, description="Human-readable reasons behind the verdict")


class ScreeningResponse(BaseModel):
    document_id: str
    ocr: OCRResult
    ela: ELAResult
    face_match: FaceMatchResult
    result: ScreeningVerdict
